### PR TITLE
[Feature] TheoreticalSpectrumGenerator speedup

### PR DESCRIPTION
**Pull request by timosachsenberg**
TheoreticalSpectrumGenerator now uses linear time computation of fragments instead of quadratic one

### CODE REVIEW

**Comment by timosachsenberg** [comment]
Failing style tests:
The following tests FAILED:
	273 - src/openms/source/FILTERING/SMOOTHING/FastLowessSmoothing.cpp_cppcheck_test (Failed)
	836 - src/utils/OpenSwathMzMLFileCacher.cpp_cppcheck_test (Failed)
are unrelated

File: src/openms/include/OpenMS/CHEMISTRY/AASequence.h [hunk 1]
```diff
@@ -376,6 +376,12 @@ namespace OpenMS
     /// returns the Id of the N-term modification; an empty string is returned if none was set
     const String& getNTerminalModification() const;
 
+    /// returns the N-term modification; nullptr is returned if none was set
+    const ResidueModification * getNTerminalResidueModification() const;
```
Comment (cbielow) [inline src/openms/include/OpenMS/CHEMISTRY/AASequence.h]: can you add a test for these two new methods?!

What about Python bindings?!

**Comment by cbielow** [comment]
looking at the code, there is some weird code,
e.g.
https://github.com/timosachsenberg/OpenMS/blob/feature/TSG_speedup/src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp#L245
The meta value is set in the first iteration, and will stay there for the whole cluster. So it can be done before the loop.

**Comment by cbielow** [comment]
same here https://github.com/timosachsenberg/OpenMS/blob/feature/TSG_speedup/src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp#L318

**Comment by cbielow** [comment]
and here
https://github.com/timosachsenberg/OpenMS/blob/feature/TSG_speedup/src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp#L481

**Comment by cbielow** [comment]
actually all occurrences of 
p.setMetaValue(...
can be optimized

**Comment by cbielow** [comment]
+1 if tests are ok/unrelated fail

**Pull request merged**